### ISSUE TITLE
torch.batch_norm_backward_reduce does not use FP32 precision for reduction under torch.amp low-precision operations as expected

### ISSUE THREAD

**Issue by Thinksky5124** [pytorch/pytorch#195701]
### 🐛 Describe the bug

The reproduction script is as follows:

Under low-precision training, `grad_output` and `input_` are in half precision, while the other inputs are in FP32. According to the `torch.amp` documentation, BatchNorm is a precision-sensitive operator and should perform reductions in FP32. When using `fp32_safe_batch_norm_backward_reduce` to simulate the computation, all values remain finite. However, calling `torch.batch_norm_backward_reduce` directly yields `inf` and `nan` results, which contradicts the expected FP32 reduction behavior.

```python
import torch


def fp32_safe_batch_norm_backward_reduce(
    grad_output, input_, mean, invstd, weight, input_g, weight_g, bias_g
):
    """Same reduction as ``torch.batch_norm_backward_reduce``, but composed
    entirely from native ops that accumulate in fp32 -- no custom kernel, no
    hat_engine, just ``.float()`` before every reduction.
    """
    dims = [0] + list(range(2, input_.dim()))
    go32 = grad_output.float()
    x32 = input_.float()
    mean32 = mean.float().view(1, -1, *([1] * (input_.dim() - 2)))
    invstd32 = invstd.float()

    sum_dy = go32.sum(dims) if input_g else None
    sum_dy_xmu = (go32 * (x32 - mean32)).sum(dims) if input_g else None
    grad_weight = (
        (sum_dy_xmu * invstd32).to(weight.dtype) if weight_g else None
    )
    grad_bias = sum_dy.to(weight.dtype) if bias_g else None
    return sum_dy, sum_dy_xmu, grad_weight, grad_bias


def main():
    assert torch.cuda.is_available(), "this repro needs a CUDA device"

    # NHW = N * H * W elements per channel; each contributes ~50 to sum_dy,
    # so NHW * 50 must clear Half's max finite value (65504) to trigger the
    # narrowing overflow, while staying trivially representable at fp16
    # granularity (no rounding noise to muddy the result).
    N, C, H, W = 8, 4, 32, 32  # NHW = 8192 -> sum_dy ~= 8192 * 50 = 409600
    device = "cuda"

    grad_output = torch.full((N, C, H, W), 50.0, device=device, dtype=torch.float16)
    input_ = torch.full((N, C, H, W), 10.0, device=device, dtype=torch.float16)
    mean = torch.zeros(C, device=device, dtype=torch.float32)
    invstd = torch.ones(C, device=device, dtype=torch.float32)
    weight = torch.ones(C, device=device, dtype=torch.float32)

    print(f"input: fp16, shape=({N},{C},{H},{W}), NHW={N * H * W} elements/channel")
    print("grad_output = 50.0 (const), input = 10.0 (const), mean = 0, invstd = 1")
    print(f"true fp32 sum_dy per channel = {N * H * W * 50.0:.1f}  "
          f"(fp16 max finite = 65504)\n")

    native_sum_dy, native_sum_dy_xmu, native_gw, native_gb = (
        torch.batch_norm_backward_reduce(
            grad_output, input_, mean, invstd, weight, True, True, True
        )
    )
    print("native torch.batch_norm_backward_reduce:")
    print(f"  sum_dy      = {native_sum_dy.tolist()}")
    print(f"  sum_dy_xmu  = {native_sum_dy_xmu.tolist()}")
    print(f"  grad_weight = {native_gw.tolist()}")
    print(f"  grad_bias   = {native_gb.tolist()}")
    print(f"  isfinite    = {torch.isfinite(native_sum_dy).all().item()}")

    fixed_sum_dy, fixed_sum_dy_xmu, fixed_gw, fixed_gb = (
        fp32_safe_batch_norm_backward_reduce(
            grad_output, input_, mean, invstd, weight, True, True, True
        )
    )
    print("\nfp32_safe_batch_norm_backward_reduce (plain native ops, no custom kernel):")
    print(f"  sum_dy      = {fixed_sum_dy.tolist()}")
    print(f"  sum_dy_xmu  = {fixed_sum_dy_xmu.tolist()}")
    print(f"  grad_weight = {fixed_gw.tolist()}")
    print(f"  grad_bias   = {fixed_gb.tolist()}")
    print(f"  isfinite    = {torch.isfinite(fixed_sum_dy).all().item()}")

    native_overflowed = torch.isinf(native_sum_dy).any().item()
    fixed_ok = torch.isfinite(fixed_sum_dy).all().item()
    print("\n" + "=" * 60)
    if native_overflowed and fixed_ok:
        print(
            "CONFIRMED: native op overflowed to inf; the plain fp32 "
            "reduction stayed finite."
        )
    else:
        print(
            "UNEXPECTED: native_overflowed="
            f"{native_overflowed}, fixed_finite={fixed_ok} "
            "-- the repro did not reproduce as expected."
        )
    print("=" * 60)


if __name__ == "__main__":
    main()
```

Script output:

```txt
input: fp16, shape=(8,4,32,32), NHW=8192 elements/channel
grad_output = 50.0 (const), input = 10.0 (const), mean = 0, invstd = 1
true fp32 sum_dy per channel = 409600.0  (fp16 max finite = 65504)

native torch.batch_norm_backward_reduce:
  sum_dy      = [inf, inf, inf, inf]
  sum_dy_xmu  = [inf, inf, inf, inf]
  grad_weight = [inf, inf, inf, inf]
  grad_bias   = [inf, inf, inf, inf]
  isfinite    = False

fp32_safe_batch_norm_backward_reduce (plain native ops, no custom kernel):
  sum_dy      = [409600.0, 409600.0, 409600.0, 409600.0]
  sum_dy_xmu  = [4096000.0, 4096000.0, 4096000.0, 4096000.0]
  grad_weight = [4096000.0, 4096000.0, 4096000.0, 4096000.0]
  grad_bias   = [409600.0, 409600.0, 409600.0, 409600.0]
  isfinite    = True

============================================================
CONFIRMED: native op overflowed to inf; the plain fp32 reduction stayed finite.
============================================================
```

After in-depth analysis, the root cause is as follows:

When the input is fp16 (half), PyTorch's design principle dictates that "inputs may be half, but intermediate accumulation should use float (`accscalar_t`) to avoid precision loss or overflow"—this is also the standard assumption for mixed-precision training under AMP. However, the ATen operator implementation does not adhere to this assumption:

```cpp
# aten/src/ATen/native/cuda/Normalization.cuh
template <typename scalar_t, typename accscalar_t>
struct Float2 {
  accscalar_t v1, v2;   // Field types are accscalar_t (float) — appears to be a "high-precision accumulator"

  Float2() {}
  Float2(scalar_t v1, scalar_t v2) : v1(v1), v2(v2) {}   // ⚠️ Constructor parameters are scalar_t (half)!
  ...
  Float2& operator+=(const Float2& a) {
    v1 += a.v1;
    v2 += a.v2;
    return *this;
  }
};
```

The problem lies in this constructor: although the fields of `Float2` are `accscalar_t` (float), intended to serve as a "high-precision accumulator," every time a new `Float2` element is constructed from raw data (at the input stage of the reduction), it invokes the `Float2(scalar_t v1, scalar_t v2)` constructor—whose parameter types are `scalar_t`, i.e., half.

In other words, before each element participating in the reduction can "enter the accumulator," it must first be implicitly converted to half, and then converted back from half to float to be stored in the `v1`/`v2` fields. The subsequent reductions within the same block/warp (via `operator+=`) do correctly perform addition in float—that part is fine. However, every newly loaded element must first be representable in half before being "absorbed" into the accumulator via this constructor. Reductions are performed hierarchically (warp shuffle → block reduce → grid reduce), and each level produces a partial sum. When this partial sum serves as an input element to the next reduction level, it too must pass through the `Float2(scalar_t, scalar_t)` constructor—meaning the partial sum itself is also narrowed down to half.

### Versions

```
PyTorch version: 2.8.0+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Rocky Linux 9.3 (Blue Onyx) (x86_64)
GCC version: (GCC) 11.4.1 20230605 (Red Hat 11.4.1-2)
Clang version: Could not collect
CMake version: version 3.27.8
Libc version: glibc-2.34

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.14.0-362.24.1.el9_3.0.1.x86_64-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.61
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 5090
GPU 1: NVIDIA GeForce RTX 5090
GPU 2: NVIDIA GeForce RTX 5090
GPU 3: NVIDIA GeForce RTX 5090
GPU 4: NVIDIA GeForce RTX 5090
GPU 5: NVIDIA GeForce RTX 5090
GPU 6: NVIDIA GeForce RTX 5090
GPU 7: NVIDIA GeForce RTX 5090

Nvidia driver version: 570.211.01
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.9.22.0
/usr/lib64/libcudnn_adv.so.9.22.0
/usr/lib64/libcudnn_cnn.so.9.22.0
/usr/lib64/libcudnn_engines_precompiled.so.9.22.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.22.0
/usr/lib64/libcudnn_engines_tensor_ir.so.9.22.0
/usr/lib64/libcudnn_ext.so.9.22.0
/usr/lib64/libcudnn_graph.so.9.22.0
/usr/lib64/libcudnn_heuristic.so.9.22.0
/usr/lib64/libcudnn_ops.so.9.22.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 9334 32-Core Processor
CPU family:                         25
Model:                              17
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        3910.2529
CPU min MHz:                        1500.0000
BogoMIPS:                           5400.19
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                     AMD-V
L1d cache:                          2 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           64 MiB (64 instances)
L3 cache:                           256 MiB (8 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-31,64-95
NUMA node1 CPU(s):                  32-63,96-127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] flake8==3.9.2
[pip3] flake8-bugbear==21.4.3
[pip3] flake8-comprehensions==3.6.0
[pip3] flake8-polyfill==1.0.2
[pip3] flake8-tidy-imports==4.5.0
[pip3] horizon-hat-engine-full==2.0.1.dev202608252030+5bb23bb.torch280.cu128
[pip3] horizon-plugin-pytorch==3.2.3.dev20260211+cu128.torch280.34a9a
[pip3] msgpack-numpy==0.4.8
[pip3] mypy==0.910
[pip3] mypy_extensions==0.4.4
[pip3] numpy==1.24.2
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.26.0
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.3
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.8.0+cu128
[pip3] torchaudio==2.8.0+cu128
[pip3] torchmetrics==0.5.0
[pip3] torchvision==0.23.0+cu128
[pip3] torchviz==0.0.3
[pip3] transformer-engine-torch==2.15.0
[pip3] triton==3.4.0
[conda] horizon-hat-engine-full               2.0.1.dev202608252030+5bb23bb.torch280.cu128  pypi_0           pypi
[conda] horizon-plugin-pytorch                3.3.3.dev20260520+cu128.torch280.4938f        dev_0            <develop>
[conda] msgpack-numpy                         0.4.8                                         pypi_0           pypi
[conda] numpy                                 1.24.2                                        pypi_0           pypi
[conda] nvidia-cublas-cu12                    12.8.4.1                                      pypi_0           pypi
[conda] nvidia-cuda-cupti-cu12                12.8.90                                       pypi_0           pypi
[conda] nvidia-cuda-nvrtc-cu12                12.8.93                                       pypi_0           pypi
[conda] nvidia-cuda-runtime-cu12              12.8.90                                       pypi_0           pypi
[conda] nvidia-cudnn-cu12                     9.10.2.21                                     pypi_0           pypi
[conda] nvidia-cudnn-frontend                 1.26.0                                        pypi_0           pypi
[conda] nvidia-cufft-cu12                     11.3.3.83                                     pypi_0           pypi
[conda] nvidia-curand-cu12                    10.3.9.90                                     pypi_0           pypi
[conda] nvidia-cusolver-cu12                  11.7.3.90                                     pypi_0           pypi
[conda] nvidia-cusparse-cu12                  12.5.8.93                                     pypi_0           pypi
[conda] nvidia-cusparselt-cu12                0.7.1                                         pypi_0           pypi
[conda] nvidia-nccl-cu12                      2.27.3                                        pypi_0           pypi
[conda] nvidia-nvjitlink-cu12                 12.8.93                                       pypi_0           pypi
[conda] nvidia-nvtx-cu12                      12.8.90                                       pypi_0           pypi
[conda] torch                                 2.8.0+cu128                                   pypi_0           pypi
[conda] torchaudio                            2.8.0+cu128                                   pypi_0           pypi
[conda] torchmetrics                          0.5.0                                         pypi_0           pypi
[conda] torchvision                           0.23.0+cu128                                  pypi_0           pypi
[conda] torchviz                              0.0.3                                         pypi_0           pypi
[conda] transformer-engine-torch              2.15.0                                        pypi_0           pypi
[conda] triton                                3.4.0                                         pypi_0           pypi
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @mcarilli @leslie-fang-intel @jgong5
